File: ios/NestMind/Sources/Models/AppModels.swift
import Foundation

struct UserProfile: Codable, Equatable {
  let userId: UUID
  let householdId: UUID?
  let householdName: String?
  let householdRole: String?
  var displayName: String
  var preferredName: String
  var assistantTone: String
  var supportStyle: String
  var lifeFocuses: [String]
  var likes: [String]
  var dislikes: [String]
  var boundaries: [String]
  var onboardingAnswers: [String: String]
  var summary: String
  var isOnboardingComplete: Bool
  let createdAt: Date?
  let updatedAt: Date?
}

struct ProfileDraft: Codable, Equatable {
  var displayName: String
  var preferredName: String
  var assistantTone: String
  var supportStyle: String
  var lifeFocusesText: String
  var likesText: String
  var dislikesText: String
  var boundariesText: String
  var onboardingAnswers: [String: String]
  var summary: String
  var isOnboardingComplete: Bool

  init(
    displayName: String,
    preferredName: String,
    assistantTone: String,
    supportStyle: String,
    lifeFocusesText: String,
    likesText: String,
    dislikesText: String,
    boundariesText: String,
    onboardingAnswers: [String: String],
    summary: String,
    isOnboardingComplete: Bool
  ) {
    self.displayName = displayName
    self.preferredName = preferredName
    self.assistantTone = assistantTone
    self.supportStyle = supportStyle
    self.lifeFocusesText = lifeFocusesText
    self.likesText = likesText
    self.dislikesText = dislikesText
    self.boundariesText = boundariesText
    self.onboardingAnswers = onboardingAnswers
    self.summary = summary
    self.isOnboardingComplete = isOnboardingComplete
  }

  static let empty = ProfileDraft(
    displayName: "",
    preferredName: "",
    assistantTone: "Warm, direct, and observant.",
    supportStyle: "Coach me gently, but do not sugarcoat the truth.",
    lifeFocusesText: "",
    likesText: "",
    dislikesText: "",
    boundariesText: "",
    onboardingAnswers: [:],
    summary: "",
    isOnboardingComplete: false
  )

  init(profile: UserProfile) {
    displayName = profile.displayName
    preferredName = profile.preferredName
    assistantTone = profile.assistantTone
    supportStyle = profile.supportStyle
    lifeFocusesText = profile.lifeFocuses.joined(separator: ", ")
    likesText = profile.likes.joined(separator: ", ")
    dislikesText = profile.dislikes.joined(separator: ", ")
    boundariesText = profile.boundaries.joined(separator: ", ")
    onboardingAnswers = profile.onboardingAnswers
    summary = profile.summary
    isOnboardingComplete = profile.isOnboardingComplete
  }

  var isReadyForOnboardingSubmission: Bool {
    !preferredName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
      !assistantTone.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
      !supportStyle.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
      onboardingAnswers.values.contains(where: { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty })
  }

  var payload: ProfileUpdateRequest {
    ProfileUpdateRequest(
      displayName: displayName.trimmedOrNil,
      preferredName: preferredName.trimmedOrNil,
      assistantTone: assistantTone.trimmedOrNil,
      supportStyle: supportStyle.trimmedOrNil,
      lifeFocuses: Self.splitCSV(lifeFocusesText),
      likes: Self.splitCSV(likesText),
      dislikes: Self.splitCSV(dislikesText),
      boundaries: Self.splitCSV(boundariesText),
      onboardingAnswers: onboardingAnswers,
      summary: summary.trimmedOrNil,
      isOnboardingComplete: isOnboardingComplete
    )
  }

  static func splitCSV(_ raw: String) -> [String] {
    raw
      .split(separator: ",")
      .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
      .filter { !$0.isEmpty }
  }
}

struct ConversationSummary: Codable, Identifiable, Equatable {
  let id: UUID
  let title: String?
  let lastMessageAt: Date?
  let createdAt: Date?
}

struct ConversationDetail: Codable, Identifiable, Equatable {
  let id: UUID
  let title: String?
  let messages: [ChatMessage]
}

struct ChatMessage: Codable, Identifiable, Equatable {
  let id: UUID
  let role: ChatRole
  let content: String
  let reasoning: String?
  let createdAt: Date?
  let attachments: [MediaAsset]
}

enum ChatRole: String, Codable, CaseIterable {
  case system
  case user
  case assistant
  case tool
}

struct MemoryEntry: Codable, Identifiable, Equatable {
  let id: UUID
  var kind: MemoryKind
  var title: String
  var summary: String
  var status: MemoryStatus
  var confidence: Double
  let createdAt: Date?
  let updatedAt: Date?
  let lastReinforcedAt: Date?
}

enum MemoryKind: String, Codable, CaseIterable, Identifiable {
  case identity
  case preference
  case goal
  case routine
  case context

  var id: String { rawValue }
}

enum MemoryStatus: String, Codable, CaseIterable, Identifiable {
  case active
  case archived
  case rejected

  var id: String { rawValue }
}

enum MemoryFeedback: String, Codable, CaseIterable, Identifiable {
  case accepted
  case rejected
  case reinforced

  var id: String { rawValue }
}

struct MediaAsset: Codable, Identifiable, Equatable {
  let id: UUID
  let contentType: String
  let analysisSummary: String?
  let storagePath: String?
  let createdAt: Date?
}

struct PendingAttachment: Identifiable, Equatable {
  let id: UUID
  let asset: MediaAsset
  let localLabel: String
}

struct ChatRequest: Encodable {
  let conversationId: UUID?
  let message: String
  let mediaAssetIds: [UUID]
}

struct ChatResponse: Decodable {
  let conversation: ConversationDetail
  let profile: UserProfile?
  let suggestedMemories: [MemoryEntry]
}

struct MediaAnalyzeRequest: Encodable {
  let conversationId: UUID?
  let prompt: String
  let imageBase64: String
  let contentType: String
}

struct MediaAnalyzeResponse: Decodable {
  let asset: MediaAsset
  let summary: String
}

struct ProfileResponse: Decodable {
  let profile: UserProfile
}

struct MemoriesResponse: Decodable {
  let memories: [MemoryEntry]
}

struct ConversationsResponse: Decodable {
  let conversations: [ConversationSummary]
}

struct ConversationResponse: Decodable {
  let conversation: ConversationDetail
}

struct ProfileUpdateRequest: Encodable {
  let displayName: String?
  let preferredName: String?
  let assistantTone: String?
  let supportStyle: String?
  let lifeFocuses: [String]
  let likes: [String]
  let dislikes: [String]
  let boundaries: [String]
  let onboardingAnswers: [String: String]
  let summary: String?
  let isOnboardingComplete: Bool
}

struct MemoryUpdateRequest: Encodable {
  let kind: MemoryKind
  let title: String
  let summary: String
  let status: MemoryStatus
}

struct MemoryFeedbackRequest: Encodable {
  let feedback: MemoryFeedback
  let note: String?
}

enum OnboardingPrompt: String, CaseIterable, Identifiable {
  case whatMattersMost = "What matters most in your life right now?"
  case currentFriction = "Where are you feeling friction or overwhelm?"
  case idealSupport = "How should the assistant support you on hard days?"
  case avoidances = "What should the assistant avoid doing or saying?"
  case dailyRhythm = "What does a good day usually look like for you?"

  var id: String { rawValue }
}

extension String {
  var trimmedOrNil: String? {
    let value = trimmingCharacters(in: .whitespacesAndNewlines)
    return value.isEmpty ? nil : value
  }
}

File: apps/android/app/src/main/java/com/nestmind/app/models/AppModels.kt
package com.nestmind.app.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// ─── User Profile ────────────────────────────────────────────────────────────

@Serializable
data class UserProfile(
    val id: String,
    @SerialName("user_id") val userId: String,
    @SerialName("household_id") val householdId: String? = null,
    @SerialName("display_name") val displayName: String = "",
    @SerialName("preferred_name") val preferredName: String = "",
    @SerialName("assistant_tone") val assistantTone: String = "",
    @SerialName("support_style") val supportStyle: String = "",
    @SerialName("life_focuses") val lifeFocuses: List<String> = emptyList(),
    val likes: List<String> = emptyList(),
    val dislikes: List<String> = emptyList(),
    val boundaries: List<String> = emptyList(),
    @SerialName("onboarding_answers") val onboardingAnswers: Map<String, String> = emptyMap(),
    val summary: String? = null,
    @SerialName("is_onboarding_complete") val isOnboardingComplete: Boolean = false,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null
)

@Serializable
data class ProfileUpdateRequest(
    @SerialName("display_name") val displayName: String? = null,
    @SerialName("preferred_name") val preferredName: String? = null,
    @SerialName("assistant_tone") val assistantTone: String? = null,
    @SerialName("support_style") val supportStyle: String? = null,
    @SerialName("life_focuses") val lifeFocuses: List<String> = emptyList(),
    val likes: List<String> = emptyList(),
    val dislikes: List<String> = emptyList(),
    val boundaries: List<String> = emptyList(),
    @SerialName("onboarding_answers") val onboardingAnswers: Map<String, String> = emptyMap(),
    val summary: String? = null,
    @SerialName("is_onboarding_complete") val isOnboardingComplete: Boolean = false
)

// ─── Conversation & Chat ─────────────────────────────────────────────────────

@Serializable
data class ConversationSummary(
    val id: String,
    val title: String? = null,
    @SerialName("last_message_at") val lastMessageAt: String? = null,
    @SerialName("created_at") val createdAt: String? = null
)

@Serializable
data class ConversationDetail(
    val id: String,
    val title: String? = null,
    val messages: List<ChatMessage> = emptyList()
)

@Serializable
data class ChatMessage(
    val id: String,
    val role: ChatRole,
    val content: String,
    val reasoning: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    val attachments: List<MediaAsset> = emptyList()
)

@Serializable
enum class ChatRole {
    @SerialName("system") SYSTEM,
    @SerialName("user") USER,
    @SerialName("assistant") ASSISTANT,
    @SerialName("tool") TOOL
}

@Serializable
data class ChatRequest(
    @SerialName("conversation_id") val conversationId: String? = null,
    val message: String,
    @SerialName("media_asset_ids") val mediaAssetIds: List<String> = emptyList()
)

@Serializable
data class ChatResponse(
    val conversation: ConversationDetail,
    val profile: UserProfile? = null,
    @SerialName("suggested_memories") val suggestedMemories: List<MemoryEntry> = emptyList()
)

// ─── Memory ──────────────────────────────────────────────────────────────────

@Serializable
data class MemoryEntry(
    val id: String,
    val kind: MemoryKind,
    val title: String,
    val summary: String,
    val status: MemoryStatus,
    val confidence: Double = 1.0,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
    @SerialName("last_reinforced_at") val lastReinforcedAt: String? = null
)

@Serializable
enum class MemoryKind {
    @SerialName("identity") IDENTITY,
    @SerialName("preference") PREFERENCE,
    @SerialName("goal") GOAL,
    @SerialName("routine") ROUTINE,
    @SerialName("context") CONTEXT
}

@Serializable
enum class MemoryStatus {
    @SerialName("active") ACTIVE,
    @SerialName("archived") ARCHIVED,
    @SerialName("rejected") REJECTED
}

@Serializable
enum class MemoryFeedback {
    @SerialName("accepted") ACCEPTED,
    @SerialName("rejected") REJECTED,
    @SerialName("reinforced") REINFORCED
}

@Serializable
data class MemoryUpdateRequest(
    val kind: MemoryKind,
    val title: String,
    val summary: String,
    val status: MemoryStatus
)

@Serializable
data class MemoryFeedbackRequest(
    val feedback: MemoryFeedback,
    val note: String? = null
)

// ─── Media ───────────────────────────────────────────────────────────────────

@Serializable
data class MediaAsset(
    val id: String,
    @SerialName("content_type") val contentType: String,
    @SerialName("analysis_summary") val analysisSummary: String? = null,
    @SerialName("storage_path") val storagePath: String? = null,
    @SerialName("created_at") val createdAt: String? = null
)

// ─── API Response Envelopes ──────────────────────────────────────────────────

@Serializable
data class ProfileResponse(val profile: UserProfile)

@Serializable
data class MemoriesResponse(val memories: List<MemoryEntry>)

@Serializable
data class ConversationsResponse(val conversations: List<ConversationSummary>)

@Serializable
data class ConversationResponse(val conversation: ConversationDetail)

@Serializable
data class ApiErrorEnvelope(val error: String)

// ─── Onboarding Prompts ───────────────────────────────────────────────────────

enum class OnboardingPrompt(val question: String) {
    WHAT_MATTERS_MOST("What matters most in your life right now?"),
    CURRENT_FRICTION("Where are you feeling friction or overwhelm?"),
    IDEAL_SUPPORT("How should the assistant support you on hard days?"),
    AVOIDANCES("What should the assistant avoid doing or saying?"),
    DAILY_RHYTHM("What does a good day usually look like for you?")
}
